### ISSUE TITLE
How to save captured images?

### ISSUE THREAD

**Issue by Nicknick203** [ultralytics/ultralytics#9165]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I cant find help anywhere for this.

I have yolov8 setup in an conda enviorment and I am trying to save a frame when a person is detected. I cannot get the program to save an image to a selected folder when it detects a person or object. I cant even get it to save anything to the folder.

I simply want to be able to save images from my yolov8 webcam video feed to a specified folder.

Current Code

from ultralytics import YOLO
import cv2

# Load a model
model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)

# Use the model
results = model.predict(source="0", show=True)

print(results)

**Comment by glenn-jocher** [comment]
@Nicknick203 hello there! 👋 Saving captured frames when a person is detected can definitely be managed with the YOLOv8 model. Here’s a concise way to save images from your webcam video feed when a person is detected:

```python
from ultralytics import YOLO
import cv2

def save_detected_frames(source=0, save_dir="detected_frames"):
    model = YOLO("yolov8n.pt")
    cap = cv2.VideoCapture(source)
    
    while True:
        ret, frame = cap.read()
        if not ret:
            break
        
        results = model.predict(frame)
        # Check for person detection (class_id for person usually is 0)
        if any(x.cls == 0 for x in results[0].boxes):
            # Construct filename with format detected_frame_<frame_number>.jpg
            filename = f"{save_dir}/detected_frame_{cap.get(cv2.CAP_PROP_POS_FRAMES):.0f}.jpg"
            cv2.imwrite(filename, frame)
            print(f"Frame saved to {filename}")

save_detected_frames()
```

Make sure that the `save_dir` exists before running the script or add code to create it if it doesn't. This script checks every frame for a person, and when it finds one, it saves the frame to the specified folder.

Feel free to adjust the `model.predict` arguments, and ensure your `yolov8n.pt` model is correctly pathed. Let me know if this helps or if you have further questions. Happy coding! 🚀

**Comment by Nicknick203** [comment]
Thankyou for your help! I was able to get images to save only when a person is detected. I do have another question, I want to be able to view a camera feed with yolov8 prediction boxes while the program is running and I would like to be able to view the prediction boxes around the saved image frames. Currently it just displayes the raw camera feed and saves raw images without prediction boxes when I run the program. Could you help me to update my code in order to have a pop up window with yolov8 prediction boxes and saves the frames with prediction boxes around them?

Here is my current code.

from ultralytics import YOLO
import cv2
import os
import time

def save_detected_frames(source=0, save_dir="C:\\Users\\Nick\\Desktop\\Detected frames", max_images=float('inf')):
    model = YOLO("yolov8n.pt")
    cap = cv2.VideoCapture(source)
    
    if not os.path.exists(save_dir):
        os.makedirs(save_dir)
        print(f"Created directory: {save_dir}")
    
    # Counter for detected frames
    detected_frame_count = 0
    
    # Initialize last_capture_time
    last_capture_time = time.time()
    
    while True:
        ret, frame = cap.read()
        if not ret:
            break
        
        results = model.predict(frame)
        # Check for person detection (class_id for person usually is 0)
        if any(x.cls == 0 for x in results[0].boxes):
            # Only capture an image once every second
            current_time = time.time()
            if current_time - last_capture_time >= 1:
                # Construct filename with format detected_frame_<frame_number>.jpg
                filename = f"{save_dir}\\detected_frame_{detected_frame_count}.jpg"
                cv2.imwrite(filename, frame)
                print(f"Frame saved to {filename}")
                
                detected_frame_count += 1
                last_capture_time = current_time
            
            # Break the loop if the maximum number of images has been reached
            if detected_frame_count >= max_images:
                break
        
        # Display camera feed
        cv2.imshow('Camera Feed', frame)
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break  # Break the loop if 'q' is pressed
    
    cap.release()
    cv2.destroyAllWindows()

save_detected_frames()

**Comment by Nicknick203** [comment]
@glenn-jocher

**Comment by glenn-jocher** [comment]
@Nicknick203 Happy to hear your initial issue was resolved! 🎉 Let's tackle your new request. To display and save the frames with YOLOv8 prediction boxes, you’ll modify the `frame` that's being displayed and saved, using the `results.show()` for live feed display and `results.save()` for frame saving, which automatically include the prediction boxes.

Here's an adapted version of your code to include those changes:

```python
from ultralytics import YOLO
import cv2
import os
import time

def save_display_frames(source=0, save_dir="C:\\Users\\Nick\\Desktop\\Detected frames", max_images=float('inf')):
    model = YOLO("yolov8n.pt")
    cap = cv2.VideoCapture(source)
    
    if not os.path.exists(save_dir):
        os.makedirs(save_dir)
        print(f"Directory created at {save_dir}")
    
    detected_frame_count = 0
    last_capture_time = time.time()

    while True:
        ret, frame = cap.read()
        if not ret:
            break

        results = model.predict(frame)
        if any(x.cls == 0 for x in results[0].boxes):
            current_time = time.time()
            if current_time - last_capture_time >= 1:
                filename = f"{save_dir}\\detected_frame_{detected_frame_count}.jpg"
                results.save(filename)
                print(f"Frame with predictions saved to {filename}")
                
                detected_frame_count += 1
                last_capture_time = current_time
            
            if detected_frame_count >= max_images:
                break

        # Instead of showing raw frame, now shows frame with predictions
        results.show()

        if cv2.waitKey(1) & 0xFF == ord('q'):
            break

    cap.release()
    cv2.destroyAllWindows()

save_display_frames()
```

Ensure your environment is set up with the needed `ultralytics` library version, and this should display the camera feed with prediction boxes and save the detected frames also with predictions around them. If you encounter any further questions or need more assistance, feel free to ask! 🚀

**Comment by Nicknick203** [comment]
I tried this implementation of the code but I still cannot seem to get it do display and save the prediction boxes. Here is my code Do you think you could help me figure out what the problem is? @glenn-jocher Here is my current code.
 (MY CODE SUBMITTED IS BEFORE CHANGES WERE MADE)


from ultralytics import YOLO
import cv2
import os
import time
import tkinter as tk
from tkinter import simpledialog

def authenticate_user():
    root = tk.Tk()
    root.withdraw()  # Hide the root window

    # Dummy username and password (replace with your actual authentication mechanism)
    valid_username = "yolov8"
    valid_password = "test"

    username = simpledialog.askstring("Username", "Enter your username:")
    password = simpledialog.askstring("Password", "Enter your password:", show="*")

    if username == valid_username and password == valid_password:
        print("Authentication successful!")
        return True
    else:
        print("Invalid username or password. Please try again.")
        return False

def save_detected_frames(source=0, save_dir="C:\\Users\\Nick\\Desktop\\Detected frames", max_images=float('inf')):
    if not authenticate_user():
        return

    model = YOLO("yolov8n.pt")
    cap = cv2.VideoCapture(source)
    
    if not os.path.exists(save_dir):
        os.makedirs(save_dir)
        print(f"Created directory: {save_dir}")
    
    # Counter for detected frames
    detected_frame_count = 0
    
    # Initialize last_capture_time
    last_capture_time = time.time()
    
    while True:
        ret, frame = cap.read()
        if not ret:
            break
        
        results = model.predict(frame)
        # Check for person detection (class_id for person usually is 0)
        if any(x.cls == 0 for x in results[0].boxes):
            # Only capture an image once every second
            current_time = time.time()
            if current_time - last_capture_time >= 1:
                # Construct filename with format detected_frame_<frame_number>.jpg
                filename = f"{save_dir}\\detected_frame_{detected_frame_count}.jpg"
                cv2.imwrite(filename, frame)
                print(f"Frame saved to {filename}")
                
                detected_frame_count += 1
                last_capture_time = current_time
            
            # Break the loop if the maximum number of images has been reached
            if detected_frame_count >= max_images:
                break
        
        # Display camera feed
        cv2.imshow('Camera Feed', frame)
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break  # Break the loop if 'q' is pressed
    
    cap.release()
    cv2.destroyAllWindows()

save_detected_frames()

**Comment by glenn-jocher** [comment]
@Nicknick203 hey there! 😊 It looks like your code is currently saving and attempting to display the original camera feed without the predictions. To get both display and save frames with YOLOv8 prediction boxes, you should utilize the `results.show()` method for display and `results.save()` for saving frames with predictions. Here's a quick adjustment to your loop for capturing frames with predictions:

```python
while True:
    ret, frame = cap.read()
    if not ret:
        break
    
    results = model.predict(frame)
    if any(x.cls == 0 for x in results[0].boxes):
        current_time = time.time()
        if current_time - last_capture_time >= 1:
            detected_frame_count += 1
            last_capture_time = current_time

            # Save frame with predictions
            save_path = f"{save_dir}\\detected_frame_{detected_frame_count}.jpg"
            results.save(save_path=[save_path])
            print(f"Frame with predictions saved to {save_path}")
        
        # Break the loop if the maximum number of images has been reached
        if detected_frame_count >= max_images:
            break

    # Display frame with predictions
    results.show()
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break  # Break the loop if 'q' is pressed
```

By using `results.show()` for real-time display and `results.save()` for saving the frames with predictions, it will fulfill your requirements for visualizing and capturing the predicted frames. Make sure your `save_dir` exists, and adjust the above code as necessary if I've missed specific details about your setup. If you still face issues, feel free to ping me again! Happy coding! 🚀

**Comment by Nicknick203** [comment]
@glenn-jocher Thankyou for your help! you have been wonderful. I plan to credit you in a Senior class group project at Georgia Southern University if that is acceptable to you. I do have another question. Currently my code is the following:

from ultralytics import YOLO
import os
import cv2
import tkinter as tk
from tkinter import filedialog
from PIL import Image, ImageTk
import time
import subprocess

class CapstoneGUI:

    def __init__(self, window, window_title):
        self.model = YOLO("yolov8n.pt")
        
        self.window = window
        self.window.title(window_title)
        
        # Prevent window from automatically rescaling
        self.window.pack_propagate(False)
        self.window.geometry("780x500")
        
        self.vid = cv2.VideoCapture(0)
        
        # Create frame to hold both webcam feed and buttons
        frame = tk.Frame(window)
        frame.pack()
        
        # Create canvas for webcam feed
        self.canvas = tk.Canvas(frame, width=self.vid.get(cv2.CAP_PROP_FRAME_WIDTH), height=self.vid.get(cv2.CAP_PROP_FRAME_HEIGHT))
        self.canvas.pack(side=tk.LEFT, pady=10)
        
        # Create frame for buttons
        button_frame = tk.Frame(frame)
        button_frame.pack(side=tk.LEFT, padx=20)
        
        self.recording = True
        
        self.recLabel = tk.Label(button_frame, text="Recording")
        self.recLabel.pack(anchor=tk.CENTER, pady=10)
        
        self.btn_pause = tk.Button(button_frame, text="Pause", width=10, command=self.pause)
        self.btn_pause.pack(anchor=tk.CENTER, pady=10)
        
        self.btn_browse = tk.Button(button_frame, text="Change\nDirectory", width=10, command=lambda: self.new_directory())
        self.btn_browse.pack(anchor=tk.CENTER, pady=10)
        
        self.btn_open = tk.Button(button_frame, text="Saved\nPictures", width=10, command=lambda: self.open_folder_in_explorer(self.save_dir))
        self.btn_open.pack(anchor=tk.CENTER, pady=10)
        
        self.btn_exit = tk.Button(button_frame, text="Exit", width=10, command=self.exit)
        self.btn_exit.pack(anchor=tk.CENTER, pady=10)
        
        # Load the pre-trained Haar Cascade face detector
        self.face_cascade = cv2.CascadeClassifier(cv2.data.haarcascades + 'haarcascade_frontalface_default.xml')
        
        # Directory to save screenshots
        self.save_dir = 'screenshots'
        
        # Create the directory if it does not exist
        if not os.path.exists(self.save_dir):
            os.makedirs(self.save_dir)
            print(f"Created directory: {self.save_dir}")
        
        # Counter for naming the screenshots
        self.detected_frame_count = 0
        
        # Timestamp of the last screenshot
        self.last_capture_time = time.time()
        
        self.update()
        
        self.window.mainloop()
        
    def update(self):
        ret, frame = self.vid.read()
        if ret:
            
            # Do not run if recording is paused
            if self.recording == True:
                # Only capture an image once every second
                current_time = time.time()
                if current_time - self.last_capture_time >= 1:
                    results = self.model.predict(frame)
                    # Check for person detection (class_id for person usually is 0)
                    
                    if any(x.cls == 0 for x in results[0].boxes):
                        # Construct filename with format detected_frame_<frame_number>.jpg
                        filename = f"{self.save_dir}\\detected_frame_{self.detected_frame_count}.jpg"
                        cv2.imwrite(filename, frame)
                        print(f"Frame saved to {filename}")
                
                        self.detected_frame_count += 1
                        self.last_capture_time = current_time
            
            #update image
            self.photo = ImageTk.PhotoImage(image=Image.fromarray(cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)))
            self.canvas.create_image(0, 0, image=self.photo, anchor=tk.NW)
        self.window.after(10, self.update)
        
    def exit(self):
        if self.vid.isOpened():
            self.vid.release()
        self.window.destroy()

    def open_folder_in_explorer(self, folder_path):
        # Ensure the folder path exists
        if not os.path.exists(folder_path):
            print("Folder does not exist.")
            return

        # Use the appropriate command based on the operating system
        if os.name == 'nt':  # For Windows
            subprocess.Popen(['explorer', folder_path], shell=True)
        elif os.name == 'posix':  # For Linux or macOS
            subprocess.Popen(['xdg-open', folder_path])
        else:
            print("Unsupported operating system.")

    def new_directory(self):
        string = self.browse_folder()
        if string:
            norm_string = string.replace('/', '\\')
            self.save_dir = norm_string
        return

    def browse_folder(self):
        folder_path = filedialog.askdirectory()
        if folder_path:
            return(folder_path)
        
    def pause(self):
        if self.recording == True:
            self.recording = False
            self.btn_pause["text"] = "Unpause"
            self.recLabel.config(text="Paused")
        else:
            self.recording = True
            self.btn_pause["text"] = "Pause"
            self.recLabel.config(text="Recording")

        

# Create a window and pass it to CapstoneGUI
root = tk.Tk()
app = CapstoneGUI(root, "Tough Shells")


Do you think you could help me figure out how to save a video recording of the instance a person is detected? maybe like 5 seconds before and after a person has been detected in the frame?

Best Wishes ~Nicholas

**Comment by glenn-jocher** [comment]
Hey Nicholas! 😄 First off, thank you for the shout-out for your class project at Georgia Southern University; that's absolutely fine with me, and I'm glad I could be of assistance!

For your current challenge, saving a video recording around the moment a person is detected (e.g., 5 seconds before and after detection) involves buffering video frames and then writing them to a file when a person is detected. Here's a brief idea of how you might implement it:

1. **Buffer Video Frames**: You could maintain a circular buffer that constantly stores the last 5 seconds of video frames.
2. **Detection Trigger**: Once a person is detected, you continue recording for an additional 5 seconds, combining these frames with the ones in your buffer.
3. **Save Video**: Save the buffered frames along with the frames following the detection into a video file.

Here's a simple example to get you started:

```python
import numpy as np
import cv2

# Example buffer size calculation: for 30 fps, 5 sec buffer = 150 frames
buffer_size = 150
frame_buffer = np.empty((buffer_size,), dtype=object)

...

def update(self):
    global frame_buffer
    ret, frame = self.vid.read()
    if ret:
        # Shift the buffer and insert the new frame
        frame_buffer[:-1] = frame_buffer[1:]
        frame_buffer[-1] = frame

        if self.recording:
            current_time = time.time()
            if current_time - self.last_capture_time >= 1:
                results = self.model.predict(frame)
                if any(x.cls == 0 for x in results[0].boxes):
                    # Here, save the buffer + next X seconds of frames as a video
                    self.save_video_from_buffer(frame)

                    self.detected_frame_count += 1
                    self.last_capture_time = current_time

...

def save_video_from_buffer(self, detection_frame):
    # Define video writer
    fourcc = cv2.VideoWriter_fourcc(*'mp4v')
    out = cv2.VideoWriter(f"{self.save_dir}\\event_{self.detected_frame_count}.mp4", fourcc, 20.0, (640, 480))
    
    # Write buffered frames
    for frame in frame_buffer:
        if frame is not None:
            out.write(frame)
    
    # Write the detection_frame and continue recording for additional X seconds (simulated here)
    out.write(detection_frame)
    # You'll need to adjust this part to record and write the additional 5 sec after detection

    out.release()
```

This example shows how to manage the buffer and hints at how you could save it into a video file, but you'd need to adjust and flesh out the recording logic for "5 seconds after detection" based on your video's FPS.

Capturing the exact "5 seconds before and after" detection dynamically can get a bit involved because you need to account for the FPS of your video and efficiently manage the buffer and ongoing recording. For a complete solution, you might explore threading to handle the buffering and file writing concurrently to your main detection loop.

That's the overview, Nicholas! Hope that sets you on the right path. Let me know if you have any questions. 🚀
